### PR TITLE
Fix IndexOutOfBounds from the highlight pass

### PR DESCRIPTION
**Pull request by mkondratek**
## Test plan
Try these examples:
- https://github.com/sourcegraph/jetbrains/issues/2110
- https://github.com/sourcegraph/jetbrains/issues/1988

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/agent/intellij_extensions/Document.kt [hunk 1]
```diff
@@ -13,6 +13,14 @@ fun Document.codyPosition(offset: Int): Position {
 }
 
 fun Document.codyRange(startOffset: Int, endOffset: Int): Range {
+  if (startOffset < 0 ||
+      startOffset > this.textLength ||
+      endOffset > this.textLength ||
+      startOffset > endOffset) {
+    throw IllegalArgumentException(
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/intellij_extensions/Document.kt:20]: Initially I wanted to bound the values here. But maybe we should not do that? We should be aware of what arguments we pass here.

File: src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt [hunk 2]
```diff
@@ -97,6 +101,12 @@ class CodyFixHighlightPass(val file: PsiFile, val editor: Editor) :
                 null
               }
             }
+            .toList()
+
+    if (protocolDiagnostics.isEmpty()) {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt:83]: it was possible that we have been sending an empty list 😅 that was completely redundant

File: src/main/kotlin/com/sourcegraph/utils/CodyFormatter.kt [hunk 3]
```diff
@@ -35,9 +36,8 @@ class CodyFormatter {
                 .createFileFromText("TEMP", file.fileType, appendedString)
 
         val codeStyleManager = CodeStyleManager.getInstance(project)
-        // This will fail if cursor < range.startOffset + completionText.length
-        // TODO change the signature of this method or at least validate the arguments
-        codeStyleManager.reformatText(psiFile, cursor, range.startOffset + completionText.length)
+        val endOffset = max(cursor, range.startOffset + completionText.length)
+        codeStyleManager.reformatText(psiFile, cursor, endOffset)
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/utils/CodyFormatter.kt:40]: Unrelated to highlight pass but I managed to get it once

File: src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt [hunk 4]
```diff
@@ -106,15 +116,18 @@ class CodyFixHighlightPass(val file: PsiFile, val editor: Editor) :
...
           val range = document.codyRange(highlight.startOffset, highlight.endOffset)
 
           if (myRangeActions.containsKey(range)) {
             continue
           }
-          val location =
-              ProtocolLocation(
-                  uri = protocolTextDocument.uri,
-                  range = document.codyRange(highlight.startOffset, highlight.endOffset))
+          val location = ProtocolLocation(uri = protocolTextDocument.uri, range = range)
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt]: Perhaps we can get rid of that whole `protocolTextDocument`?
It seems we need it only for uri, and we can get is using `ProtocolTextDocument::uriFor`.
It is not marked as requiring EDT so there is a good chance we will also not need that future, etc..

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt]: btw, doCollectInformation is called from the background thread by IntelliJ

image
(we are extending HighlightPass and collectInformation calls doCollectInformation)

File: src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt [hunk 5]
```diff
@@ -106,15 +116,18 @@ class CodyFixHighlightPass(val file: PsiFile, val editor: Editor) :
           if (progress.isCanceled) {
             break
           }
+          if (highlight.startOffset > document.textLength ||
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt]: We are now reading document in outside of the EDT.
Your check probably highly decreases chance of the issue, but I doubt it eliminates it as state can change even between this and the next instruction.
I think `document.codyRange` should be called on EDT.

File: src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt (hunk 4 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt]: good point about the URI - fixing!

File: src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt (hunk 5 above)
Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt]: fixed

File: src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt [hunk 6]
```diff
@@ -137,6 +131,12 @@ class CodyFixHighlightPass(val file: PsiFile, val editor: Editor) :
     for (highlight in myHighlights) {
       highlight.unregisterQuickFix { it.familyName == CodeActionQuickFix.FAMILY_NAME }
 
+      if (highlight.startOffset > document.textLength ||
+          highlight.endOffset > document.textLength ||
+          highlight.startOffset > highlight.endOffset) {
+        break
+      }
+
       val range = document.codyRange(highlight.startOffset, highlight.endOffset)
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt:142]: Is that running in the EDT?

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/cody/inspections/CodyFixHighlightPass.kt:142]: It runs in EDT. I added `@RequiresEdt` on it.

**Review by pkukielka — APPROVED** [review]
LGTM

**Pull request merged**